File: src/main/java/com/attilapalf/exceptional/services/ExceptionService.java
package com.attilapalf.exceptional.services;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.attilapalf.exceptional.entities.ExceptionInstancesEntity;
import com.attilapalf.exceptional.entities.UsersEntity;
import com.attilapalf.exceptional.messages.BaseExceptionRequest;
import com.attilapalf.exceptional.messages.ExceptionInstanceWrapper;
import com.attilapalf.exceptional.messages.ExceptionRefreshResponse;
import com.attilapalf.exceptional.messages.ExceptionSentResponse;
import com.attilapalf.exceptional.repositories.exceptioninstances_.ExceptionInstanceCrud;
import com.attilapalf.exceptional.repositories.users.UserCrud;

/**
 * Created by Attila on 2015-06-21.
 */
@SuppressWarnings( "SpringJavaAutowiringInspection" )
@Service
public class ExceptionService {
    @Autowired
    private ExceptionInstanceCrud exceptionCrud;
    @Autowired
    private UserCrud userCrud;
    @Autowired
    private GcmMessageService gcmMessageService;


    @Transactional
    public ExceptionSentResponse throwException( ExceptionInstanceWrapper exceptionInstanceWrapper ) {
        UsersEntity sender = userCrud.findOne( exceptionInstanceWrapper.getFromWho() );
        UsersEntity receiver = userCrud.findOne( exceptionInstanceWrapper.getToWho() );
        updateUsersPoints( sender, receiver );
        ExceptionInstancesEntity exception = exceptionCrud.saveNewException( exceptionInstanceWrapper );
        exceptionInstanceWrapper.setInstanceId( exception.getId() );
        gcmMessageService.sendExceptionNotification( receiver, sender, exception );
        return createExceptionSentResponse( sender, receiver, exceptionInstanceWrapper, exception );
    }

    private void updateUsersPoints( UsersEntity sender, UsersEntity receiver ) {
        sender.setPoints( sender.getPoints() + 25 );
        receiver.setPoints( receiver.getPoints() - 20 );
        userCrud.save( sender );
        userCrud.save( receiver );
    }

    private ExceptionSentResponse createExceptionSentResponse( UsersEntity sender, UsersEntity receiver,
                                                               ExceptionInstanceWrapper instanceWrapper,
                                                               ExceptionInstancesEntity exception ) {
        return new ExceptionSentResponse(
                instanceWrapper,
                exception.getType().getShortName(),
                sender.getPoints(),
                receiver.getPoints() );
    }

    @Transactional
    public ExceptionRefreshResponse refreshExceptions( BaseExceptionRequest requestBody ) {
        UsersEntity user = userCrud.findOne( requestBody.getUserFacebookId() );
        List<ExceptionInstancesEntity> exceptions = exceptionCrud.findLastExceptionsNotAmongIds(
                user,
                requestBody.getKnownExceptionIds()
        );
        return new ExceptionRefreshResponse( exceptions );
    }
}

File: src/main/java/com/attilapalf/exceptional/services/ExceptionService.kt
package com.attilapalf.exceptional.services

import com.attilapalf.exceptional.entities.ExceptionInstance
import com.attilapalf.exceptional.entities.User
import com.attilapalf.exceptional.messages.*
import com.attilapalf.exceptional.repositories.exceptioninstances_.ExceptionInstanceCrud
import com.attilapalf.exceptional.repositories.users.UserCrud
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * Created by palfi on 2015-10-02.
 */
public interface ExceptionService {
    fun throwException(instanceWrapper: ExceptionInstanceWrapper): ExceptionSentResponse
    fun refreshExceptions(requestBody: BaseExceptionRequest): ExceptionRefreshResponse
    fun answerQuestion(questionAnswer: QuestionAnswer): ExceptionSentResponse
}

@Service
public class ExceptionServiceImpl : ExceptionService {
    @Autowired
    private lateinit val exceptionCrud: ExceptionInstanceCrud
    @Autowired
    private lateinit val userCrud: UserCrud
    @Autowired
    private lateinit val gcmMessageService: GcmMessageService


    @Transactional
    override fun throwException(instanceWrapper: ExceptionInstanceWrapper): ExceptionSentResponse {
        val users = findUsers(instanceWrapper)
        val exception = updateDatabase(instanceWrapper, users)
        instanceWrapper.instanceId = exception.id
        gcmMessageService.sendExceptionNotification(users.sender, users.receiver, exception, instanceWrapper.question)
        return ExceptionSentResponse(exception.type.shortName, users.sender.points, users.receiver.points, instanceWrapper)
    }

    private fun findUsers(instanceWrapper: ExceptionInstanceWrapper) =
            Users(
                    sender = userCrud.findOne(instanceWrapper.fromWho),
                    receiver = userCrud.findOne(instanceWrapper.toWho)
            )

    private fun updateDatabase(instanceWrapper: ExceptionInstanceWrapper, users: Users): ExceptionInstance {
        calculatePoints(instanceWrapper)
        updateUsersPoints(users.sender, users.receiver, instanceWrapper)
        val exception = exceptionCrud.saveNewException(instanceWrapper)
        return exception
    }

    private fun calculatePoints(instanceWrapper: ExceptionInstanceWrapper) {
        if ( !instanceWrapper.question.hasQuestion ) {
            instanceWrapper.pointsForSender = 25
            instanceWrapper.pointsForReceiver = -20
        } else {
            instanceWrapper.pointsForSender = 0
            instanceWrapper.pointsForReceiver = 0
        }
    }

    private fun updateUsersPoints(sender: User, receiver: User, instanceWrapper: ExceptionInstanceWrapper) {
        sender.points = sender.points + instanceWrapper.pointsForSender
        receiver.points = receiver.points + instanceWrapper.pointsForReceiver
        userCrud.save(sender)
        userCrud.save(receiver)
    }

    @Transactional
    override fun refreshExceptions(requestBody: BaseExceptionRequest): ExceptionRefreshResponse {
        val user = userCrud.findOne(requestBody.userFacebookId)
        val exceptions = exceptionCrud.findLastExceptionsNotAmongIds(
                user,
                requestBody.knownExceptionIds)
        return ExceptionRefreshResponse(exceptions.map { ExceptionInstanceWrapper(it) })
    }

    @Transactional
    override fun answerQuestion(questionAnswer: QuestionAnswer): ExceptionSentResponse {
        val exception = exceptionCrud.findOne(questionAnswer.exceptionInstanceId)
        updatePoints(exception, questionAnswer)
        exception.isAnswered = true
        exceptionCrud.save(exception)
        return ExceptionSentResponse(exception.type.shortName, exception.fromUser.points, exception.toUser.points,
                ExceptionInstanceWrapper(exception))
    }

    private fun updatePoints(exception: ExceptionInstance, questionAnswer: QuestionAnswer) {
        if ( exception.isYesIsCorrect != questionAnswer.answeredYes ) {
            exception.pointsForSender = 50;
            exception.pointsForReceiver = -40;
            exception.fromUser.points += exception.pointsForSender;
            exception.toUser.points += exception.pointsForReceiver;
        }
    }

    private data class Users(val sender: User, val receiver: User)
}